Create tutorial to get ZTF light curves from HATS catalog (#325)
Caltech-IPAC/irsa-tutorials: tutorials/ztf/ztf_lightcurves.md

---
authors:
- name: Jaladh Singhal
- name: Troy Raen
- name: "Brigitta Sip\u0151cz"
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.19.3
kernelspec:
  display_name: irsa-tutorials
  language: python
  name: python3
---

(ztf-lightcurves)=
# Access ZTF DR24 Light Curves from HATS Catalog

+++

## Learning Goals

By the end of this tutorial, you will learn how to:

- Open ZTF DR24 HATS catalogs for light curves and the Objects Table using `lsdb`.
- Retrieve light curves for specific sources by ZTF object IDs using an index search.
- Retrieve light curves for sources in a sky region using a cone search.

> - Cross-reference the Objects Table to enrich cone search results with per-source variability statistics.

Reviewer: I think we should use either "object" or "target" instead of "source".

```suggestion
- Retrieve light curves for specific objects by ZTF object IDs using an index search.
- Retrieve light curves for objects in a sky region using a cone search.
- Cross-reference the Objects Table to enrich cone search results with per-object variability statistics.
```

- Plot ZTF light curves (filtered by variability statistics).

+++

## Introduction

The ZTF DR24 enhanced data products at IRSA include two [HATS](https://irsa.ipac.caltech.edu/docs/parquet_catalogs/#hats) (Hierarchical Adaptive Tiling Scheme) catalogs hosted on AWS S3:

- **Lightcurves catalog**: one row per ZTF object, with a nested column storing the full photometry time series — timestamps, magnitudes, uncertainties, and quality flags.

> - **Objects Table**: one row per ZTF object per band, with collapsed light curve metrics such as magnitude RMS, chi-squared variability statistic, number of good observations, and mean magnitude.

Reviewer: Lightcurves and Objects Table contain the same objects (well, Lightcurves contains a subset), so adding "per band" to one but not the other is a bit confusing. Also, ZTF objects are actually defined per (band, field, quadrant) -- not only per band. I think we should leave out "per band" here, but add a statement about it below. (See my comment below.)

```suggestion
- **Lightcurves catalog**: one row per ZTF object, with a nested column storing the full photometry time series — timestamps, magnitudes, uncertainties, and quality flags.
- **Objects Table**: one row per ZTF object, with collapsed light curve metrics such as magnitude RMS, chi-squared variability statistic, number of good observations, and mean magnitude.
 ```

The [lsdb](https://docs.lsdb.io/en/latest/index.html) Python library provides a convenient interface for working with HATS catalogs, including spatial queries and object-ID-based lookups.

This tutorial covers two common entry points for accessing ZTF light curves:

1. **Object IDs**: you have specific ZTF object IDs — from a previous query, a catalog crossmatch, or a published source list — and want their light curves directly.
2. **RA/Dec**: you have sky coordinates and want all ZTF sources within a given radius.

Both approaches are demonstrated below. An optional section then shows how to join the position search results with the Objects Table to select and plot the most variable sources using robust variability statistics.

For more context on ZTF DR24 data products, refer to the [ZTF DR24 release notes](https://irsa.ipac.caltech.edu/data/ZTF/docs/releases/ztf_release_notes_latest) and [explanatory supplement](https://irsa.ipac.caltech.edu/data/ZTF/docs/ztf_explanatory_supplement.pdf) at IRSA.

+++

## Imports

```{code-cell} ipython3
# Uncomment the next line to install dependencies if needed.

> # !pip install s3fs "lsdb>=0.6.6,<0.8" pyarrow pandas astropy matplotlib

Reviewer: Is `lsdb<0.8` just for our own CI and due to other notebooks? I wonder if/how we can handle that without implying to end users that this particular notebook requires lsdb<0.8.

Author: Yes it's just because of our CI and other notebooks. I was planning to do #197 soon so that should remove the upper pin from this notebook.

We can also hold off merging (or at least advertising) this tutorial until #197 is closed if you think it is important?

Reviewer: FYI @bsipocz, @jaladh-singhal noticed (very) incorrect results returned by lsdb cone search and we tracked it to a version issue. Upgrading to lsdb>=0.8 fixes it. I double checked the HATS catalogs and the data are fine. Calculating separation using astropy shows that cone search returns objects that are well outside the cone. Not sure what's going on there. It's possible we're missing a crucial kwarg or something, but our upper pin of v0.7.3 is two major versions old now so better to spend our time trying to unpin. @jaladh-singhal will open a separate PR for that.

Author: Opened https://github.com/Caltech-IPAC/irsa-tutorials/pull/331 - we need to wait for that to be merged before this one is merged.

Author: Rebased and updated it to lsdb>=0.8.1

```

```{code-cell} ipython3
import s3fs
import lsdb
import pyarrow.parquet as pq
from astropy.coordinates import SkyCoord
import pandas as pd
from astropy import units as u
import os
import matplotlib.pyplot as plt
from dask.distributed import Client
```

```{code-cell} ipython3
pd.set_option("display.max_colwidth", None)
pd.set_option("display.min_rows", 18)
```

## 1. Locate ZTF DR24 HATS Catalogs in the Cloud

From IRSA's [cloud data access page](https://irsa.ipac.caltech.edu/cloud_access/), we identify the S3 bucket and path prefixes for the ZTF DR24 HATS catalogs:

```{code-cell} ipython3
ztf_bucket = "ipac-irsa-ztf"
ztf_lc_hats_prefix = "ztf/enhanced/dr24/lc/hats" # Light curves catalog

> ztf_objects_hats_prefix = "ztf/enhanced/dr24/objects/hats" # Objects table

Reviewer: ```suggestion
ztf_lc_hats_prefix = "ztf/enhanced/dr24/lc/hats" # Lightcurves catalog
ztf_objects_hats_prefix = "ztf/enhanced/dr24/objects/hats" # Objects Table
```

Use the proper nouns. (ZTF named these products "Lightcurves" and "Objects Table".)

```

[s3fs](https://s3fs.readthedocs.io/en/latest/) provides a filesystem-like Python interface for AWS S3 buckets.
First, we create an S3 client:

```{code-cell} ipython3
s3 = s3fs.S3FileSystem(anon=True)
```

> Let's list the contents of the ZTF DR24 lightcurves HATS **collection**:

Reviewer: ```suggestion
Let's list the contents of the ZTF DR24 Lightcurves HATS **Collection**:
```

Proper noun.

```{code-cell} ipython3
s3.ls(f"{ztf_bucket}/{ztf_objects_hats_prefix}")
```

```{code-cell} ipython3
ztf_objects_schema_path = "ztf_dr24_objects-hats/dataset/_common_metadata"  # ztf_dr24_objects-hats is the catalog name identified above
```

## 2. Explore the Catalog Schema

Before querying the catalogs, let's inspect what columns are available in each.
We read schemas from the `_common_metadata` files, which also contain column metadata such as units and descriptions:

```{code-cell} ipython3
def pq_schema_to_df(schema):
    """Convert a PyArrow schema to a Pandas DataFrame."""
    return pd.DataFrame(
        [
            (
                field.name,
                str(field.type),
                field.metadata.get(b"unit", b"").decode(),
                field.metadata.get(b"description", b"").decode()
            )
            for field in schema
        ],
        columns=["name", "type", "unit", "description"]
    )
```

```{code-cell} ipython3
ztf_lc_schema = pq.read_schema(
    f"s3://{ztf_bucket}/{ztf_lc_hats_prefix}/{ztf_lc_schema_path}",
    filesystem=s3
)
ztf_lc_schema_df = pq_schema_to_df(ztf_lc_schema)
ztf_lc_schema_df
```

> Notice the `lightcurve` column — this is a **nested column** that stores the full photometric time series for each ZTF object.

Reviewer: ```suggestion
Notice the `lightcurve` column — this is a **[nested](https://nested-pandas.readthedocs.io/) column** that stores the full photometric time series for each ZTF object.
```

Maybe link here. I find myself needing to refer to those docs to figure out/remember how to work with nested columns.

Each element of `lightcurve` is itself a table with columns including `hmjd`, `mag`,`magerr`, `clrcoeff` and `catflags`.
We save the list of columns interesting to us for later use when opening the catalog with `lsdb`:

```{code-cell} ipython3

> ztf_lc_columns = ["objectid", "objra", "objdec", "filterid", "nepochs", "lightcurve"]

Reviewer: I hoped we could save more memory by selecting only the `lightcurve` columns that actually get used, but I tried this and the savings is insignificant (we need 4 out of the 5 columns, so I suppose that makes sense). Leaving this suggestion in case you want to show that it's possible to load only some of the nested columns. Also fine to ignore this.

```suggestion
ztf_lc_columns = ["objectid", "objra", "objdec", "filterid", "nepochs", 
                  "lightcurve.hmjd", "lightcurve.mag", "lightcurve.magerr", "lightcurve.catflags"]
```

Author: Yes I first tried the destructed columns thinking the same but observed no performance gain. So I switched to "lightcurve" to keep it simple.

)
ztf_lc_catalog
```

### 3.2 Identify the Index Column

The ZTF DR24 light curves HATS catalog ships with an ancillary index table that enables fast lookups by object ID.
Let's identify which column is indexed:

```{code-cell} ipython3
ztf_lc_idx_column = list(ztf_lc_catalog.hc_collection.all_indexes.keys())[0]
print(f"Index column: {ztf_lc_idx_column}")
```

### 3.3 Perform an Index Search

We use the same object IDs from the [ZTF light curve API docs](https://irsa.ipac.caltech.edu/docs/program_interface/ztf_lightcurve_api.html) multi-object example — you can compare results from this tutorial directly with that service.
In your workflow, these IDs might come from a previous query, a catalog crossmatch, or a published source table:

```{code-cell} ipython3
object_ids = [686103400034440, 686103400106565]
```

```{code-cell} ipython3
ztf_lcs_by_id = ztf_lc_catalog.id_search(values={ztf_lc_idx_column: object_ids})
ztf_lcs_by_id
```

### 3.4 Compute and Inspect the Results

Now we execute the query we planned in previous steps by calling `compute()`. This is where the data is read into memory as a Pandas DataFrame.

```{code-cell} ipython3
ztf_lcs_by_id_df = ztf_lcs_by_id.compute()

> ztf_lcs_by_id_df

Reviewer: I ran this on Fornax to check memory usage and found that Dask holds onto the memory it grabs here. Adding `del ztf_lcs_by_id_df` and similar for the catalog objects doesn't help. Wrapping this in a client context does help. (same as you figured out for sec. 4 below)

11.5G = Max memory usage of notebook without `Client` here
7.1G   = Max memory usage of notebook with `Client` here

I used this:

```suggestion
def get_nworkers(object_ids):
    return min(os.cpu_count(), len(object_ids))

with Client(n_workers=get_nworkers(object_ids),
            threads_per_worker=1,
            memory_limit=None  # each partition can be several GB; avoid per-worker cap
            ) as client:
    print(f"You can monitor progress in the Dask dashboard at {client.dashboard_link}")
    ztf_lcs_by_id_df = ztf_lcs_by_id.compute()

ztf_lcs_by_id_df
```

Author: Thanks for analyzing the difference. I added it but I should note that in some cases (at least locally), I see id search compute taking more time when wrapped in a dask client. But the time increase is not significantly high to offset the memory optimization so it should be fine.

```{code-cell} ipython3
fig, axs = plt.subplots(len(ztf_lcs_by_id_df), 1,
                        figsize=(10, 4 * len(ztf_lcs_by_id_df)),
                        constrained_layout=True)

if len(ztf_lcs_by_id_df) == 1:
    axs = [axs]

for ax, (_, row) in zip(axs, ztf_lcs_by_id_df.iterrows()):
    lc = row['lightcurve'].query("catflags == 0")  # to keep only clean epochs
    title = f"ZTF Object {row['objectid']}  (RA={row['objra']:.4f}°, Dec={row['objdec']:.4f}°)"
    pts = ax.plot(lc['hmjd'], lc['mag'], '.', markersize=4, zorder=3)
    ax.errorbar(
        lc['hmjd'], lc['mag'], yerr=lc['magerr'],
        fmt='none', ecolor=pts[0].get_color(), elinewidth=0.8, alpha=0.3, zorder=2
    )
    ax.set_ylabel("Magnitude")
    ax.set_xlabel("HMJD")
    ax.invert_yaxis()
    ax.set_title(title, fontsize=10)

fig.suptitle("ZTF DR24 Light Curves — Object ID Search Results", fontsize=13, y=1.02)
plt.show()
```

## 4. Get Light Curves by Sky Position

If you have sky coordinates and want all ZTF sources within a given area, use a cone search.

> 

Reviewer: Let's add this admonition (or similar) to describe how ZTF objects are defined. This is the first place where it's directly relevant, but you could move it up to the introduction if you prefer.

```suggestion

:::{important} ZTF objects are defined per (filter, field, quadrant)
ZTF objects (i.e., unique object IDs) are defined _per_ (filter, field, quadrant).
This means that observations of a single _astrophysical_ object are usually spread out amongst several different _ZTF_ objects.

At minimum, a given astrophysical object will be represented by up to 3 ZTF objects, one per filter (g, r, and i).
The per-filter observations may themselves be separated into additional ZTF objects if the astrophysical object lies near the boundary of a ZTF field and/or quadrant.

ZTF's pixel scale is 1"/pixel (see [ZTF Technical Specifications](https://www.ptf.caltech.edu/page/ztf_technical)), so combining all ZTF objects within a 1" cone search may be reasonable for a given astrophysical object.
:::

```

You can choose whether you want to change the actual code below here or not. I think it's fine to leave as-is as long as we provide this admonition. If changing, I would probably reduce `search_radius` to 1 arcsec, remove the `["nepochs", ">", 50]` row filter, and group by filter (ie, band) before plotting the light curves. (FWIW, our Fornax light-curve-collector notebook justifies the 1" cone search by [citing Graham et al., 2024](https://github.com/nasa-fornax/fornax-demo-notebooks/blob/37370b5c9463a851abaa98f61eae0d97b16abaf0/light_curves/light_curve_collector.md?plain=1#L336). Unfortunately, ADS returns 79 papers for "Graham et al., 2024" and I don't know which one that came from.)

Author: Added this as admonition. I reduced search cone to 1" but kept row filter on "nepochs" and reduced it to >25 so as to demonstrate row filters can be applied before compute (which can be helpful in big spatial searches unlike the tutorial example).

Here we keep only objects with more than 50 epochs, focusing on well-sampled light curves:

```{code-cell} ipython3
row_filters = [
    ["nepochs", ">", 50],
    # additional filters can be added here if desired
    ]
```

### 4.3 Open the Filtered Light Curves Catalog

We open the catalog with both filters applied. lsdb evaluates this lazily — no data is read yet:

```{code-cell} ipython3
ztf_lc_cone = lsdb.open_catalog(
    f"s3://{ztf_bucket}/{ztf_lc_hats_prefix}",
    search_filter=spatial_filter,
    columns=ztf_lc_columns,
    filters=row_filters
)
ztf_lc_cone
```

Notice that only the partitions overlapping the cone are included, avoiding reads of the full catalog.

### 4.4 Compute and Inspect the Results

Now we execute the query by calling `compute()`. The ZTF DR24 LC catalog stores full nested light curves per HATS partition. We wrap the compute call in a Dask client to parallelize if multiple partitions are involved, and to monitor progress in the Dask dashboard.

```{code-cell} ipython3
def get_nworkers(catalog):
    return min(os.cpu_count(), catalog.npartitions + 1)

with Client(n_workers=get_nworkers(ztf_lc_cone),
            threads_per_worker=1,

>             memory_limit=None  # each partition can be several GB; avoid per-worker cap

Author: It was running out of memory locally without this. Let's see how it performs on CI.

            ) as client:
    print(f"You can monitor progress in the Dask dashboard at {client.dashboard_link}")
    ztf_lc_cone_df = ztf_lc_cone.compute()
```

```{code-cell} ipython3
print(f"Found {len(ztf_lc_cone_df)} ZTF light curves for the search criteria.")
```

```{code-cell} ipython3
ztf_lc_cone_df.head(5)
```

Each row corresponds to one ZTF object. The `lightcurve` column contains a nested DataFrame per object:

```{code-cell} ipython3
ztf_lc_cone_df['lightcurve'].iloc[0]
```

## 5. [Optional] Look Up Additional Info from the Objects Table

```{note}
This section is optional — skip it if you don't need additional information beyond the raw light curves from section 4.
```

### 5.1 Explore the Objects Table Schema

> The Objects Table contains per-band summary statistics for each ZTF source.

Reviewer: ```suggestion
The Objects Table contains summary statistics for each ZTF object.
```

Let's inspect its schema to identify columns of interest:

```{code-cell} ipython3
ztf_objects_schema = pq.read_schema(
    f"s3://{ztf_bucket}/{ztf_objects_hats_prefix}/{ztf_objects_schema_path}",
    filesystem=s3
)
pq_schema_to_df(ztf_objects_schema)
```

We'll select a subset of columns useful for characterizing and annotating variable sources:

> 

Reviewer: This is another thing that trips people up, so let's add an admonition. Could move this up to the intro (or elsewhere) if you prefer.

```suggestion

:::{important} `objectid` == `oid`
ZTF's object ID column is named `objectid` in Lightcurves and `oid` in Objects Table.
Despite this difference, the two columns are the same and can be used to join the catalogs. 
:::

```

ztf_objects_columns = ['oid', 'ra', 'dec', 'filtercode', 'ngoodobsrel', 'chisq', 'magrms', 'meanmag', 'medianabsdev']
```

### 5.2 Open the Objects Table

We reuse the same `spatial_filter` from section 4 to retrieve Objects Table entries for the same sky region. This is important for ensuring we only retrieve rows relevant to the light curves we got from the position search.

```{code-cell} ipython3
ztf_objects_cone = lsdb.open_catalog(
    f"s3://{ztf_bucket}/{ztf_objects_hats_prefix}",
    search_filter=spatial_filter,
    columns=ztf_objects_columns
)
ztf_objects_cone
```

### 5.3 Compute and Inspect

```{code-cell} ipython3
with Client(n_workers=get_nworkers(ztf_objects_cone),
            threads_per_worker=1,
            memory_limit=None) as client:
    ztf_objects_cone_df = ztf_objects_cone.compute()
```

```{code-cell} ipython3
ztf_objects_cone_df
```

### 5.4 Merge Objects Table Info into Light Curves

We merge the Objects Table with the position search light curves on the shared object ID via an inner join:

```{code-cell} ipython3
combined_df = ztf_lc_cone_df.merge(
    ztf_objects_cone_df,
    left_on='objectid',
    right_on='oid',
    how='inner'
)

> combined_df

Author: Just FYI, I tried 3 more approaches before settling on this one (which takes takes 2m±5s both compute calls combined):
1. `ztf_objects_cone.join(ztf_lc_cone, ...).compute()` takes 2m±5s
2. `ztf_objects_cone.merge(ztf_lc_cone, ...).compute()` takes 2m±5s
3. `ztf_objects_cone.compute(); ztf_lc_cone.id_search(values={'objectid': list(ztf_objects_cone_df.oid)}).compute()` took 2m50s

I kept this approach because of time as well as to maintain the narrative of keeping objects table search optional.

```

## 6. Plot Most Variable Light Curves from the Position Search

Using the `chisq` column, we rudimentarily select the top 3 most variable sources from the position search results combined with objects table.

```{code-cell} ipython3
# most_variable = ztf_lc_cone_df  # uncomment if you skipped section 5, and comment the line below
most_variable = combined_df.nlargest(3, 'chisq')
most_variable
```

Then we plot their light curves annotated with summary statistics:

```{code-cell} ipython3
fig, axs = plt.subplots(len(most_variable), 1,
                        figsize=(10, 4 * len(most_variable)),
                        constrained_layout=True)

if len(most_variable) == 1:
    axs = [axs]

for ax, (_, row) in zip(axs, most_variable.iterrows()):
    lc = row['lightcurve'].query("catflags == 0")  # to keep only clean epochs

>     title = (f"ZTF Object {row['objectid']}  ({row['filtercode']} band)\n"

Reviewer: Maybe we should stick with "filter". Most people will understand that band == filter but newbies may be confused if we don't explain it.

```suggestion
    title = (f"ZTF Object {row['objectid']}  ({row['filtercode']} filter)\n"
```

---
authors:
- name: Jaladh Singhal
- name: Troy Raen
- name: "Brigitta Sip\u0151cz"
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.19.3
kernelspec:
  display_name: irsa-tutorials
  language: python
  name: python3
---

(ztf-lightcurves)=
# Access ZTF DR24 Light Curves from HATS Catalog

+++

## Learning Goals

By the end of this tutorial, you will learn how to:

> - Open ZTF DR24 HATS catalogs for light curves and the Objects Table using `lsdb`.

Reviewer: I stumbled over this sentence and wonder if something like this would be clearer.

```suggestion
- Open ZTF DR24 Lightcurves and Objects Table HATS Catalogs using `lsdb`.
```

```{code-cell} ipython3
fig, axs = plt.subplots(len(ztf_lcs_by_id_df), 1,
                        figsize=(10, 4 * len(ztf_lcs_by_id_df)),
                        constrained_layout=True)

if len(ztf_lcs_by_id_df) == 1:
    axs = [axs]

for ax, (_, row) in zip(axs, ztf_lcs_by_id_df.iterrows()):
    lc = row['lightcurve'].query("catflags == 0")  # to keep only clean epochs
    title = f"ZTF Object {row['objectid']}  (RA={row['objra']:.4f}°, Dec={row['objdec']:.4f}°)"
    pts = ax.plot(lc['hmjd'], lc['mag'], '.', markersize=4, zorder=3)
    ax.errorbar(
        lc['hmjd'], lc['mag'], yerr=lc['magerr'],
        fmt='none', ecolor=pts[0].get_color(), elinewidth=0.8, alpha=0.3, zorder=2
    )
    ax.set_ylabel("Magnitude")
    ax.set_xlabel("HMJD")
    ax.invert_yaxis()
    ax.set_title(title, fontsize=10)

fig.suptitle("ZTF DR24 Light Curves — Object ID Search Results", fontsize=13, y=1.02)
plt.show()
```

## 4. Get Light Curves by Sky Position

> If you have sky coordinates and want all ZTF objects within a given area, use a cone search.

Reviewer: Let's point out that cone search isn't the only option.

```suggestion
If you have sky coordinates and want all ZTF objects within a given area, use a [spatial search](https://docs.lsdb.io/en/latest/tutorials/region_selection.html).
We'll do a cone search here.
Box and polygon searches are also available.
```

            memory_limit=None  # each partition can be several GB; avoid per-worker cap
            ) as client:
    print(f"You can monitor progress in the Dask dashboard at {client.dashboard_link}")
    ztf_lc_cone_df = ztf_lc_cone.compute()
```

```{code-cell} ipython3
print(f"Found {len(ztf_lc_cone_df)} ZTF light curves for the search criteria.")
```

```{code-cell} ipython3
ztf_lc_cone_df.head(5)
```

Each row corresponds to one ZTF object. The `lightcurve` column contains a nested DataFrame per object:

```{code-cell} ipython3
ztf_lc_cone_df['lightcurve'].iloc[0]
```

## 5. [Optional] Look Up Additional Info from the Objects Table

```{note}
This section is optional — skip it if you don't need additional information beyond the raw light curves from section 4.
```

### 5.1 Explore the Objects Table Schema

The Objects Table contains summary statistics for each ZTF object.
Let's inspect its schema to identify columns of interest:

```{code-cell} ipython3
ztf_objects_schema = pq.read_schema(
    f"s3://{ztf_bucket}/{ztf_objects_hats_prefix}/{ztf_objects_schema_path}",
    filesystem=s3
)
pq_schema_to_df(ztf_objects_schema)
```

```{important} objectid == oid

> ZTF's object ID column is named `objectid` in Lightcurves and `oid` in Objects Table.
> Despite this difference, the two columns are the same and can be used to join the catalogs. 
> ```
> 
> We'll select a subset of columns useful for characterizing and annotating variable objects:
> 
> ```{code-cell} ipython3
> ztf_objects_columns = ['oid', 'ra', 'dec', 'filtercode', 'ngoodobsrel', 'chisq', 'magrms', 'meanmag', 'medianabsdev']
> ```
> 
> ### 5.2 Open the Objects Table
> 
> We reuse the same `spatial_filter` from section 4 to retrieve Objects Table entries for the same sky region. This is important for ensuring we only retrieve rows relevant to the light curves we got from the position search.

Reviewer: One could also take the object IDs from the light curves results and use them to do an index search on the Objects Table, right? I don't have a good sense of when to recommend one over the other; maybe you do? (Index search used to be quite slow but I think it has improved?) Also, I remembered another thing that trips people up sometimes and deserves a note here...

````suggestion
### 5.2 Open the Objects Table

As with Lightcurves, we can load objects from the Objects Table by doing either an index search or a spatial search.
We'll do a spatial search here.

```{note}
Objects Table contains more objects than Lightcurves.
This is because objects were identified using co-added reference images that have deeper photometry than the single-exposure detection limits.
So objects may be "missing" from Lightcurves, or may be present but sparsely sampled, when their photometry is fainter or approximately equal to the single-exposure detection limits.
See item 9 in the "What you should know about the ZTF Release Products" section of the [release notes](https://irsa.ipac.caltech.edu/data/ZTF/docs/releases/ztf_release_notes_latest).
```

We reuse the same `spatial_filter` from section 4 to retrieve Objects Table entries for the same sky region. This is important for ensuring we only retrieve rows relevant to the light curves we got from the position search.
````

Author: ID search (at least in my experience) is slower and this section builds upon extending the light curves results with objects data so I went with just spatial search. But it doesn't hurt to add both are possible. Thanks for the writeup along with the note.

Reviewer: Feel free to give more direct advice/guidance/insight, maybe "We'll do a spatial search here, which is generally faster.", or whatever you think is best. Users definitely appreciate advice on which to choose when there are two options that accomplish the same thing.
